pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,391 @@
+// directory.  Actual files are stored in a hidden timestamped directory which
+// is symlinked to by the data directory. The timestamped directory and
+// data directory symlink are created in the writer's target dir.  This scheme
+// allows the files to be atomically updated by changing the target of the
+// data directory symlink.
+type AtomicWriter struct {
+	TargetDir  string
+	LogContext string
+}
+
+func NewAtomicWriter(targetDir, podNamespace, podName, volumeName string) (*AtomicWriter, error) {
+	_, err := os.Stat(targetDir)
+	if os.IsNotExist(err) {
+		return nil, err
+	}
+
+	context := fmt.Sprintf("%v/%v volume: %v", podNamespace, podName, volumeName)
+	return &AtomicWriter{TargetDir: targetDir, LogContext: context}, nil
+}
+
+const (
+	sentinelFileName = "..sentinel"
+	dataDirName      = "..data"
+	newDataDirName   = "..data_tmp"
+)
+
+// Write does an atomic projection of the given payload into the writer's target
+// directory. Note, AtomicWriter reserves the set of pathnames containing `..`.
+// Input files should not contain this path; Write will return an error if they do.
+// Write may add or delete user-visible files in the target directory.
+//
+// The Write algorithm is:
+//
+//  0.  The payload is validated; if the payload is invalid, the function returns
+//  1.  A check is made to determine whether data present in the payload has changed
+//  2.  The user-visible portion of the volume is walked to determine whether any
+//      portion of the payload was deleted and is still present on disk.
+//      If the payload is already present on disk and there are no deleted files,
+//      the function returns
+//  3.  A new timestamped dir is created
+//  4.  The payload is written to the new timestamped directory
+//  5.  Symlinks and directory for new user-visible files are created (if needed).
+//
+//      For example consider the files:
+//        <target-dir>/podName
+//        <target-dir>/user/labels
+//        <target-dir>/k8s/annotations
+//
+//      The user visible files are symbolic links into the internal data directory:
+//        <target-dir>/podName         -> ..data/podName
+//        <target-dir>/usr/labels      -> ../..data/usr/labels
+//        <target-dir>/k8s/annotations -> ../..data/k8s/annotations
+//
+//      Relative links are created into the data directory for files in subdirectories.
+//
+//      The data directory itself is a link to a timestamped directory with
+//      the real data:
+//        <target-dir>/..data          -> ..data.12345678/
+//  6.  The current timestamped directory is detected by reading the data directory
+//      symlink
+//  7.  A symlink to the new timestamped directory ..data_tmp is created that will
+//      become the new data directory
+//  8.  The new data directory symlink is renamed to the data directory; rename is atomic
+//      unless the system goes down during it, in which case it will present as a hard-link
```

Comment:
I'll remove it, I have seen it written a couple places that if the system goes down during rename, it will present as a hard-link.  Either way, it's too much detail for this comment, will strike it.